8300517: Refactor VisibleMemberTable (method members)

pavelrappo opened the pull request:
Please review this explorative refactoring for VisibleMemberTable (VMT).

This is the first round of refactoring for VMT. This round is about *method members*: declared (overriding and not) and inherited.

During this work I gained some insight into internal workings of VMT, fixed what was feasible and left TODOs and FIXMEs for everything else. Leaving those comments might look untidy, but leaving them out is wasteful: they clearly mark issues that should be revisited in upcoming rounds of refactoring.

As I see it today, the main issue with VMT is that implements complex and error-prone computations from Java Language Specification (JLS) by hand. For example, VMT interprets JLS rules for relations such as _inherits_, _overrides_ and _hides_. As one would imagine, sometimes VMT does it incorrectly. It would be better to eventually re-implement VMT using `javax.lang.model` as much as possible. Unlike that of `jdk.javadoc`, the day job of `javax.lang.model` is to provide JLS services.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8300517](https://bugs.openjdk.org/browse/JDK-8300517): Refactor VisibleMemberTable (method members)


### Reviewers
 * [Jonathan Gibbons](https://openjdk.org/census#jjg) (@jonathan-gibbons - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12887/head:pull/12887` \
`$ git checkout pull/12887`

Update a local copy of the PR: \
`$ git checkout pull/12887` \
`$ git pull https://git.openjdk.org/jdk pull/12887/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12887`

View PR using the GUI difftool: \
`$ git pr show -t 12887`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12887.diff">https://git.openjdk.org/jdk/pull/12887.diff</a>

</details>

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/doclet/package-info.java:
```diff
@@ -79,10 +79,13 @@
- *       <em>specified</em> if it contains a <em>specified</em> element,
- *       or it is enclosed in a <em>specified</em> element, and is <em>selected</em>.
- *       Included elements will be documented.
+ *   <dd>An element is considered to be <em>included</em>, if it is <em>selected</em> and any of the following it true:
```
typo: following **is** true

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/doclet/package-info.java:
```diff
@@ -79,10 +79,13 @@
+ *   <dd>An element is considered to be <em>included</em>, if it is <em>selected</em> and any of the following it true:
+ *     <ul>
+ *       <li>the element is <em>specified</em>, or
+ *       <li>the element contains a <em>specified</em> element, or
```
for clarification: I presume this is recursive/inductive ... i.e this implies all enclosing elements are specified

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/HtmlDocletWriter.java:
```diff
@@ -289,23 +289,30 @@ public String replaceDocRootDir(String htmlstr) {
-            }
+        var enclosing = (TypeElement) method.getEnclosingElement();
+        var overrideInfo = utils.overriddenMethod(method);
+        var vmt = configuration.getVisibleMemberTable(enclosing);
```
minor style suggestion, since you like clear informative name ...
Calling it just `vmt` makes it seem like a singleton;  how about calling it `enclosingVmt`? (I agree the camelCase is horrible)

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/HtmlDocletWriter.java:
```diff
@@ -289,23 +289,30 @@ public String replaceDocRootDir(String htmlstr) {
+            //   "Specified by" documentation. The examples of that can be seen in documentation
+            //   for these methods:
+            //     * ForkJoinPool.execute(java.lang.Runnable)
+            //  This is a long-standing bug, which must be fixed separately.
```
Recommend include a JBS issue to track this.

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/MethodWriterImpl.java:
```diff
@@ -293,26 +294,28 @@ protected static void addOverridden(HtmlDocletWriter writer,
                                             Content dl) {
         Utils utils = writer.utils;
-        if (utils.isStatic(method) || writer.options.noComment()) {
+        if (writer.options.noComment()) {
```
I presume that `static` methods are filtered beforehand

pavelrappo replied:
Correct; this is ensured by the only call site: HtmlDocletWriter.addMethodInfo

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/MethodWriterImpl.java:
```diff
@@ -293,26 +294,28 @@ protected static void addOverridden(HtmlDocletWriter writer,
         Contents contents = writer.contents;
-        VisibleMemberTable vmt = writer.configuration
-                .getVisibleMemberTable(utils.getEnclosingTypeElement(method));
+        var enclosing = (TypeElement) method.getEnclosingElement();
```
Wishlist for Language Model: default methods that return a narrower result type:
```
interface Demo {
   ...>     Element getEnclosingElement(Element e);
   ...>     default TypeElement getEnclosingElement(ExecutableElement ee) {
   ...>         return (TypeElement) getEnclosingElement((Element) ee);
   ...>     }
   ...>     default TypeElement getEnclosingElement(VariableElement ve) {
   ...>         return (TypeElement) getEnclosingElement((Element) ve);
   ...>     }
   ...> }
   ```

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/WorkArounds.java:
```diff
@@ -194,74 +194,6 @@ public TypeElement searchClass(TypeElement klass, String className) {
-               rider.overrides(ridee, origin, javacTypes, true);
-    }
-
     // TODO: jx.l.m ?
```
Not edited, but I note that Language Model and Compiler API have a very distant relationship :-(

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/Utils.java:
```diff
@@ -333,8 +276,19 @@ public boolean isFinal(Element e) {
+     * This is to avoid confusion with unrelated _default_ methods which
+     * appeared in JDK 8.
+     */
     public boolean isPackagePrivate(Element e) {
```
Maybe change the method name, in line with the comment?

pavelrappo replied:
I would refrain from doing it in the context of this PR.

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/Utils.java:
```diff
@@ -660,7 +610,15 @@ private boolean shouldInstantiate(TypeElement site, Element e) {
     /*
+     * The record is used to pass the method along with the type where that method is visible.
+     * Passing the type explicitly allows to preserve a complete type information, including
+     * parametrization, which is otherwise unavailable without computation similar to what
```
typo: parameterization

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/Utils.java:
```diff
@@ -673,43 +631,29 @@ private boolean shouldInstantiate(TypeElement site, Element e) {
+            var te = (TypeElement) ((DeclaredType) t).asElement();
+            assert te.getKind().isClass();
             VisibleMemberTable vmt = configuration.getVisibleMemberTable(te);
             for (Element e : vmt.getMembers(VisibleMemberTable.Kind.METHODS)) {
```
In similar situations, I have found it useful to pass in a type token that determines the result type and which avoids casting the returned objects:

For example:
```
   <T> List<T extends Element> getMembers(VisibleMemberTable.Kind k, Class<T> clazz)
```

pavelrappo replied:
Unfortunately, there's no such thing as generic enum, which would be just the ticket here.

Type token would be okay, but perhaps only marginally better than what we have now: neither option supports static checking. Anyhow, we might revisit it in the next round of refactoring, not now.

jonathan-gibbons replied:
OK, with not now, but I do think the type token is marginally better than a cast of the result at the use site.

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/Utils.java:
```diff
@@ -1093,28 +1026,23 @@ public TypeElement getFirstVisibleSuperClassAsTypeElement(TypeElement te) {
+            }
+            t = supertypes.get(0); // if non-empty, the first element is always the superclass
+            var te = asTypeElement(t);
+            assert alreadySeen.add(te);
```
I guess it's OK, but it is a mild uugh to see an imperative method as an expression for `assert`

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/Utils.java:
```diff
@@ -1093,28 +1026,23 @@ public TypeElement getFirstVisibleSuperClassAsTypeElement(TypeElement te) {
-            return null;
+        // TODO: this computation should be eventually delegated to VisibleMemberTable
+        Set<TypeElement> alreadySeen = null;
+        assert (alreadySeen = new HashSet<>()) != null; // create set conditionally
```
I think this use of `assert` is a step too far.  It's one thing to use asserts to verify invariants; it's too much to use them to conditionally compute state like this.

pavelrappo replied:
That hash map is needed iff assertions are enabled. For an ordinary run, creating that hash map could be wasteful.

Perhaps I could create a hash map unconditionally using `HashSet.newHashSet(0)`; would that be to your liking?

jonathan-gibbons replied:
The issue was more about the composite declaration and use of the asserts to build a collection solely for use in the assertions. This is a level of assert above simply checking invariants with no other side effects on the JVM.  The use is clever and not wrong, but it is uncommon.  Maybe a higher level comment might help, more than the low level "create set conditionally"

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/VisibleMemberTable.java:
```diff
@@ -261,17 +281,26 @@ public List<Element> getMembers(Kind kind) {
-     * primarily used to determine the location of a possible comment.
+     * Returns the method overridden by the provided method, or {@code null}.
+     *
+     * Sometimes it's not possible to link to a method that a link, linkplain,
+     * or see tag mentions. This is because the method is a "simple override"
+     * and, thus, has no useful documentation, or because the method is
+     * declared in a type that has package access and, thus, has no visible
+     * documentation.
+     *
```
Nice comment

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/VisibleMemberTable.java:
```diff
@@ -261,17 +281,26 @@ public List<Element> getMembers(Kind kind) {
         ensureInitialized();
-
-        OverriddenMethodInfo found = overriddenMethodTable.get(e);
+        assert !overriddenMethodTable.containsKey(null);
```
can we somehow do the assert at the time of `put` ?

jonathan-gibbons replied:
Understood, but this is more about the general pattern for handling (any) bad values in (any) collection.

Yes, you can check the values when you retrieve them, but then all you know is that _someone_ added/put a bad value, but not _who_ or _when_. In general, it is better to detect that bad value as early as possible so that you still have the surrounding context of _who_ and _when_ to help determine _why_.

pavelrappo replied:
Let's not complicate it now; that code will go away pretty soon anyway.

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/VisibleMemberTable.java:
```diff
@@ -141,23 +141,44 @@ public static Set<Kind> forDetailsOf(ElementKind kind) {
+     * Tables for the superinterfaces might not be unique (i.e. an interface
+     * may be added from different lineages).
+     */
     private final List<VisibleMemberTable> parents;
```
Should this be something different, such as a Set or a multi-map?
Generally, javac and to some extent javadoc have historically used lists for "list-y sets".

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/VisibleMemberTable.java:
```diff
@@ -516,75 +545,127 @@ private void computeVisibleFieldsAndInnerClasses(LocalMemberTable lmt, Kind kind
-    boolean isEnclosureInterface(Element e) {
-        TypeElement enclosing = utils.getEnclosingTypeElement(e);
-        return utils.isPlainInterface(enclosing);
+    private static <K, V> boolean putAllIsNonReplacing(Map<K, V> dst, Map<K, V> src) {
```
naming suggestion:   the dominant initial word here is `put`  but the method is a predicate, suggesting that the initial word should be `is`

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/VisibleMemberTable.java:
```diff
@@ -516,75 +545,127 @@ private void computeVisibleFieldsAndInnerClasses(LocalMemberTable lmt, Kind kind
-                                 Map<ExecutableElement, List<ExecutableElement>> overriddenByTable,
-                                 LocalMemberTable lmt) {
-        if (!isInherited(inheritedMethod))
+    private boolean allowInheritedMethod(ExecutableElement inheritedMethod,
```
what does `allow` mean in this context? if it's inherited, doesn't that mean it's allowed? how can an inherited method not be allowed?

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/VisibleMemberTable.java:
```diff
@@ -610,11 +691,12 @@ boolean allowInheritedMethod(ExecutableElement inheritedMethod,
             ExecutableElement lMethod = (ExecutableElement) le;
             // Ignore private methods or those methods marked with
-            // a "hidden" tag.
+            // a "hidden" tag. // FIXME I cannot see where @hidden is ignored
```
Generally, there is `utils.hasHiddenTag`

It _may_ be a good design question as to when `@hidden` should be taken into account.  For example, should VMT build an accurate model and only suppress on output, or should `@hidden` be reflected in the VMT model.

If it helps (?) I note that at one time a precursor of `@hidden` was `@treatAsPrivate`

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/Utils.java:
```diff
@@ -660,7 +610,15 @@ private boolean shouldInstantiate(TypeElement site, Element e) {
+                               DeclaredType overriddenMethodOwner) { }
+
     /*
+     * The record is used to pass the method along with the type where that method is visible.
+     * Passing the type explicitly allows to preserve a complete type information, including
+     * parameterization, which is otherwise unavailable without computation similar to what
+     * this method does.
+     *
```
Consider moving this part of the comment onto the comment for the record declaration itself.

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/Utils.java:
```diff
@@ -1093,28 +1026,23 @@ public TypeElement getFirstVisibleSuperClassAsTypeElement(TypeElement te) {
      *          be found.
```
Although not edited, this comment could be improved by defining the term "closest visible".
In particular, I notice the implementation includes reference to `@hidden` tags.

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/Utils.java:
```diff
@@ -1093,28 +1026,23 @@ public TypeElement getFirstVisibleSuperClassAsTypeElement(TypeElement te) {
-        }
-        if (typeUtils.isSameType(type, superType)) {
-            return null;
+        // TODO: this computation should be eventually delegated to VisibleMemberTable
```
Agreed, both the computation and this method and it's `TypeElement` sibling.

jonathan-gibbons reviewed (approved):
Approved, although there are still some outstanding suggestions for comments.

Generally, this is obviously a big project, so well done on progress so far.

It is well worth aligning the functionality and terminology of this code with current JLS, and so I (generally) encourage JLS references in key places, preferably in a stylized manner (to facilitate checking and/or updates).

I also think it is worth adding notes comparing the functionality here with that of plain-old `Elements.getAllMembers`.  That is a simple building block that returns a list-y set of the members of a `TypeElement`, but VMT needs to do more:
* handle members of so-called hidden types
* build a DAG for each method, relating that method to corresponding methods in the supertypes of the type element, for use by `@link` and `Specified By:`.
* ...

pavelrappo commented:
@jonathan-gibbons, while I haven't addressed all your review comments, I believe I addressed the most important ones.

Unaddressed comments will either be explicitly addressed in upcoming PRs or will become irrelevant as a result of code change in those PRs.

pavelrappo commented:
/integrate
